File: CS/TreeListFilter/FilterTreeList/ColumnFilter/Operation.cs
using System.Collections.Generic;
using DevExpress.XtraTreeList.Columns;
using DevExpress.XtraTreeList.Nodes;
using DevExpress.XtraTreeList.Nodes.Operations;

namespace FilterTreeListControl
{
	public class NodesValuesOperation : TreeListOperation
	{
		private readonly TreeListColumn column;
		private readonly List<ColumnFilterItem> result = new List<ColumnFilterItem>();

		public NodesValuesOperation(TreeListColumn column)
		{
			this.column = column;
		}

		public override void Execute(TreeListNode node)
		{
			result.Add(new ColumnFilterItem(node, column));
		}

		public List<ColumnFilterItem> Result
		{
			get { return result; }
		}
	}

	public class FilterNodesOperation : TreeListOperation
	{
		private readonly ColumnFilterConditionCollection filterConditions;
		private readonly List<TreeListNode> filteredVisibleNodes = new List<TreeListNode>();

		public FilterNodesOperation(ColumnFilterConditionCollection filterConditions)
		{
			this.filterConditions = filterConditions;
		}

		private bool IsNodeFiltered(TreeListNode node)
		{
			int metConditions = 0;
			foreach ( ColumnFilterCondition condition in filterConditions )
			{
				object value = node.GetValue(condition.Column);
				if ( condition.CheckValue(value) )
					metConditions++;
			}

			if ( metConditions == filterConditions.Count )
				return false;

			return true;
		}

		public override void Execute(TreeListNode node)
		{
			bool visibleChildNodesResult = false;
			if ( node.Nodes.Count != 0 )
			{
				FilterNodesOperation childNodesOperation = new FilterNodesOperation(filterConditions);
				node.TreeList.NodesIterator.DoLocalOperation(childNodesOperation, node.Nodes);
				visibleChildNodesResult = childNodesOperation.HasVisibleChildNodes;
				filteredVisibleNodes.AddRange(childNodesOperation.FilteredVisibleNodes);
			}

			bool nodeFiltered = IsNodeFiltered(node);
			if ( !nodeFiltered || visibleChildNodesResult )
				hasVisibleChildNodes = true;

			if ( nodeFiltered && visibleChildNodesResult )
				filteredVisibleNodes.Add(node);

			node.Visible = !nodeFiltered || visibleChildNodesResult;
		}

		public override bool NeedsVisitChildren(TreeListNode node)
		{
			return false;
		}

		private bool hasVisibleChildNodes;
		public bool HasVisibleChildNodes
		{
			get { return hasVisibleChildNodes; }
		}
		
		public List<TreeListNode> FilteredVisibleNodes
		{
			get { return filteredVisibleNodes; }
		}
	}
}

File: VB/TreeListFilter/FilterTreeList/ColumnFilter/Operation.vb
Imports System.Collections.Generic
Imports DevExpress.XtraTreeList.Columns
Imports DevExpress.XtraTreeList.Nodes
Imports DevExpress.XtraTreeList.Nodes.Operations

Namespace FilterTreeListControl
	Public Class NodesValuesOperation
		Inherits TreeListOperation
		Private ReadOnly column As TreeListColumn
		Private ReadOnly resultList As New List(Of ColumnFilterItem)()

		Public Sub New(ByVal column As TreeListColumn)
			Me.column = column
		End Sub

		Public Overrides Sub Execute(ByVal node As TreeListNode)
			resultList.Add(New ColumnFilterItem(node, column))
		End Sub

		Public ReadOnly Property Result() As List(Of ColumnFilterItem)
			Get
				Return resultList
			End Get
		End Property
	End Class

	Public Class FilterNodesOperation
		Inherits TreeListOperation
		Private ReadOnly filterConditions As ColumnFilterConditionCollection
		Private ReadOnly _filteredVisibleNodes As New List(Of TreeListNode)()

		Public Sub New(ByVal filterConditions As ColumnFilterConditionCollection)
			Me.filterConditions = filterConditions
		End Sub

		Private Function IsNodeFiltered(ByVal node As TreeListNode) As Boolean
			Dim metConditions As Integer = 0
			For Each condition As ColumnFilterCondition In filterConditions
				Dim value As Object = node.GetValue(condition.Column)
				If condition.CheckValue(value) Then
					metConditions += 1
				End If
			Next condition

			If metConditions = filterConditions.Count Then
				Return False
			End If

			Return True
		End Function

		Public Overrides Sub Execute(ByVal node As TreeListNode)
			Dim visibleChildNodesResult As Boolean = False
			If node.Nodes.Count <> 0 Then
				Dim childNodesOperation As New FilterNodesOperation(filterConditions)
				node.TreeList.NodesIterator.DoLocalOperation(childNodesOperation, node.Nodes)
				visibleChildNodesResult = childNodesOperation.HasVisibleChildNodes
				_filteredVisibleNodes.AddRange(childNodesOperation.FilteredVisibleNodes)
			End If

			Dim nodeFiltered As Boolean = IsNodeFiltered(node)
			If (Not nodeFiltered) OrElse visibleChildNodesResult Then
				_hasVisibleChildNodes = True
			End If

			If nodeFiltered AndAlso visibleChildNodesResult Then
				_filteredVisibleNodes.Add(node)
			End If

			node.Visible = (Not nodeFiltered) OrElse visibleChildNodesResult
		End Sub

		Public Overrides Function NeedsVisitChildren(ByVal node As TreeListNode) As Boolean
			Return False
		End Function

		Private _hasVisibleChildNodes As Boolean
		Public ReadOnly Property HasVisibleChildNodes() As Boolean
			Get
				Return _hasVisibleChildNodes
			End Get
		End Property

		Public ReadOnly Property FilteredVisibleNodes() As List(Of TreeListNode)
			Get
				Return _filteredVisibleNodes
			End Get
		End Property
	End Class
End Namespace
